src/eval.rs

```diff
@@ -117,23 +117,130 @@ pub enum ThunkState {
 /// The mutable data stored inside a thunk.
 #[derive(Clone, Debug, PartialEq)]
 pub struct ThunkData {
-    closure: Closure,
+    inner: InnerThunkData,
     state: ThunkState,
 }
 
+/// The part of [ThunkData] responsible for storing the closure itself. It can either be:
+/// - A standard thunk, that is destructively updated once and for all
+/// - A reversible thunk, that can be restored to its original expression. Used to implement
+/// recursive merging of records and overriding (see the [RFC
+/// overriding](https://github.com/tweag/nickel/pull/330))
+#[derive(Clone, Debug, PartialEq)]
+pub enum InnerThunkData {
+    Standard(Closure),
+    Reversible {
+        orig: Rc<Closure>,
+        cached: Option<Closure>,
+    },
+}
+
 impl ThunkData {
+    /// Create new standard thunk data.
     pub fn new(closure: Closure) -> Self {
         ThunkData {
-            closure,
+            inner: InnerThunkData::Standard(closure),
             state: ThunkState::Suspended,
         }
     }
+
+    /// Create new reversible thunk data.
+    pub fn new_rev(orig: Closure) -> Self {
+        ThunkData {
+            inner: InnerThunkData::Reversible {
+                orig: Rc::new(orig),
+                cached: None,
+            },
+            state: ThunkState::Suspended,
+        }
+    }
+
+    /// Return a reference to the closure currently cached.
+    pub fn closure(&self) -> &Closure {
+        match self.inner {
+            InnerThunkData::Standard(ref closure) => closure,
+            InnerThunkData::Reversible {
+                ref orig,
+                cached: None,
+            } => orig,
+            InnerThunkData::Reversible {
+                cached: Some(ref closure),
+                ..
+            } => closure,
+        }
+    }
+
+    /// Return a mutable reference to the closure currently cached.
+    pub fn closure_mut(&mut self) -> &mut Closure {
+        match self.inner {
+            InnerThunkData::Standard(ref mut closure) => closure,
+            InnerThunkData::Reversible {
+                ref orig,
+                ref mut cached,
+            } => {
+                if cached.is_none() {
+                    *cached = Some((**orig).clone());
+                }
+                cached.as_mut().unwrap()
+            }
+        }
+    }
+
+    /// Consume the data and return the cached closure.
+    pub fn into_closure(self) -> Closure {
+        match self.inner {
+            InnerThunkData::Standard(closure) => closure,
+            InnerThunkData::Reversible { orig, cached: None } => {
+                Rc::try_unwrap(orig).unwrap_or_else(|rc| (*rc).clone())
```

Comment:
I don't even think that's the case. Because we have ownership over `self`, we can explicitly `std::mem::drop(orig)` before trying to unwrap `cached`. This should avoid cloning in the same conditions: when this thunk is only one to refer to the rc-ed expression.